### PR TITLE
Application Screenshots included

### PR DESCRIPTION
**Pull request by traliman**
New property IncludeScreenshots in ISettings is made. New class ScreenShotWriter is made, ReportStorage is  modified to include screenshots in zip file. Examples is  modified to see the local copies of zip files(LocalSender class).

### CODE REVIEW

File: Examples/ConsoleExample/LocalSender.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.IO;
+using NCrash.Core;
+using NCrash.Sender;
+
+namespace NCrash.Examples.ConsoleExample
+{
+    class localsender : ISender
```
Comment (akurdyukov) [inline Examples/ConsoleExample/LocalSender.cs]: Naming rules violation. Class name should start with capital letter. This class should be 'LocalSender'.

File: Examples/ConsoleExample/LocalSender.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        FileStream filew;
+        byte[] output;
+        public bool Send(Stream data, string fileName, Report report)
+        {
+
+            filew = new FileStream(fileName, FileMode.Create);
```
Comment (akurdyukov) [inline Examples/ConsoleExample/LocalSender.cs]: This code will cause problems if something between FileStream creation and Close throws exception. Need 'using' here.

File: Examples/ConsoleExample/Program.cs [hunk 3]
```diff
@@ -10,7 +10,13 @@ static void Main()
         {
             var userInterface = new EmptyUserInterface {Flow = ExecutionFlow.BreakExecution};
             var settings = new DefaultSettings {HandleProcessCorruptedStateExceptions = true, UserInterface = userInterface};
+            //mycode
+            settings.Sender = new localsender();
+            settings.IncludeScreenshots = true;
             var reporter = new ErrorReporter(settings);
+            reporter.HandleExceptions = true;
+            //System.IO.IsolatedStorage.IsolatedStorageFile.Remove(System.IO.IsolatedStorage.IsolatedStorageScope.User);
+            //mycode
```
Comment (akurdyukov) [inline Examples/ConsoleExample/Program.cs]: Cannot merge debug comments.

File: Examples/WinFormsExample/LocalSender.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.IO;
+using NCrash.Core;
+using NCrash.Sender;
+
+namespace NCrash.Examples.WinFormsExample
+{
+    class localsender : ISender
```
Comment (akurdyukov) [inline Examples/WinFormsExample/LocalSender.cs]: Same comments as above.

File: NCrash/Core/ScreenShots/ScreenShotWriter.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
+﻿//mycode
```
Comment (akurdyukov) [inline NCrash/Core/ScreenShots/ScreenShotWriter.cs]: This comment is not useful

File: NCrash/Core/ScreenShots/ScreenShotWriter.cs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        [StructLayout(LayoutKind.Sequential)]
+        private struct Rect
+        {
+            public int Left;
+            public int Top;
+            public int Right;
+            public int Bottom;
+        }
+
+        private static List<int> Windows = new List<int>();
```
Comment (akurdyukov) [inline NCrash/Core/ScreenShots/ScreenShotWriter.cs]: invalid naming, should me '_windows'. See other classes for style.
Also, too wide type, you can create IList here

File: NCrash/Core/ScreenShots/ScreenShotWriter.cs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            try
+            {
+                EnumWindows(new WindowEnumCallback(AddWnd), 0);
+                screenshotNum = 1;
+                screenshotNames.Clear();
+                foreach (int hwnd in Windows)
+                {
+
+                    screenshot = Capture(hwnd);
+                    screenshotName = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData),
```
Comment (akurdyukov) [inline NCrash/Core/ScreenShots/ScreenShotWriter.cs]: This path should be tunable

File: NCrash/Core/ScreenShots/ScreenShotWriter.cs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+                                                    "Application_screenshot_" + screenshotNum.ToString() + ".jpg"); ;
+                    screenshot.Save(screenshotName);
+                    screenshotNum++;
+                    screenshotNames.Add(screenshotName);
+                }
+            }
+            catch (Exception e)
```
Comment (akurdyukov) [inline NCrash/Core/ScreenShots/ScreenShotWriter.cs]: Very bad idea to catch exception, but show nothing.

File: NCrash/Storage/ReportStorage.cs [hunk 9]
```diff
@@ -54,6 +55,22 @@ public void Write(Report report)
                         {
                             AddAdditionalFiles(zipStorer);
                         }
+                        
+                        //mycode
+                        // Include application screenshots (if configured so)
+                        if (_settings.IncludeScreenshots)
+                        {
+                            int i = 1;
+                            System.Collections.Generic.List<string> names =  ScreenShotWriter.Write();
```
Comment (akurdyukov) [inline NCrash/Storage/ReportStorage.cs]: Why use full class reference?

**Comment by traliman** [comment]
Added unserialized property -  the list of screenshot files to report. Screenshots are generated in ErrorReporter.Report(). The user can attach them to zip file in  ProcessingException event handler. Added the example how to do this.

File: Examples/ConsoleExample/Program.cs [hunk 10]
```diff
@@ -10,7 +12,19 @@ static void Main()
...
             var settings = new DefaultSettings {HandleProcessCorruptedStateExceptions = true, UserInterface = userInterface};
+            settings.Sender = new LocalSender();
             var reporter = new ErrorReporter(settings);
+            reporter.HandleExceptions = true;
+            /// Example how to add screenshots, or other files to report
+            reporter.ProcessingException += (ex, report) =>
+            {
+                if (settings.AdditionalReportFiles == null)
+                    settings.AdditionalReportFiles = new List<string>();
+                foreach (Tuple<string, string> screenshot in report.ScreenshotList)
```
Comment (akurdyukov) [inline Examples/ConsoleExample/Program.cs]: why this adding is here, not inside ErrorReporter class?

File: NCrash/Core/ScreenShots/ScreenShotWriter.cs [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+            GetWindowThreadProcessId(hwnd, out windowProcessId);
+            if (currentProcessId == windowProcessId && IsWindowVisible(hwnd))
+                _windows.Add(hwnd);
+            return true;
+        }
+        ///<summary>
+        ///Geneating screenshots
+        /// </summary>
+
+        internal static IList<Tuple<string,string>> Write(string _path)
```
Comment (akurdyukov) [inline NCrash/Core/ScreenShots/ScreenShotWriter.cs]: Underscore (_) is for private fields, not arguments.

File: NCrash/ErrorReporter.cs [hunk 12]
```diff
@@ -267,6 +268,8 @@ protected virtual ExecutionFlow Report(Exception exception)
                 Logger.Trace("Starting to generate a bug report for the exception.");
                 var serializableException = new SerializableException(exception);
                 var report = new Report(serializableException);
+                //Generating screenshots, adding list of result files to report.
+                report.ScreenshotList = ScreenShotWriter.Write(string.Empty);
```
Comment (akurdyukov) [inline NCrash/ErrorReporter.cs]: Why create special fields and not use standard AdditionalFiles? You still have screenshot-specific code in generic class.

File: NCrash/Core/Report.cs [hunk 13]
```diff
@@ -39,5 +41,11 @@ public override string ToString()
         {
             return string.Format("{{CustomInfo: {0}, GeneralInfo: {1}, Exception: {2}}}", CustomInfo, GeneralInfo, Exception);
         }
+
+        /// <summary>
+        /// Unserialized property. List of screenshot files. Path,file.
+        /// </summary>
+        [XmlIgnoreAttribute]
+        public IList<Tuple<string, string>> ScreenshotList{get;set;}
```
Comment (akurdyukov) [inline NCrash/Core/Report.cs]: See code formatting in "{get;set;}". Does not match other code style.

**Comment by traliman** [comment]
Plugins added. Processing and postprocessing plugins in ErrorReporter.Report. ScreenShotWriter plugin released.

File: NCrash/ISettings.cs [hunk 14]
```diff
@@ -37,7 +38,7 @@ public interface ISettings
         /// Gets or sets a list of additional files to be added to the report zip. The files can use * or ? in the same way as DOS modifiers.
         /// Paths should be absolute i.e. start with disk name.
         /// </summary>
-        IList<string> AdditionalReportFiles { get; }
+        IList<string> AdditionalReportFiles { get; set; }
```
Comment (akurdyukov) [inline NCrash/ISettings.cs]: Why user setter here?

File: NCrash/Plugins/IPlugin.cs [hunk 15]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace NCrash.Plugins
+{
+    public interface IPlugin
```
Comment (akurdyukov) [inline NCrash/Plugins/IPlugin.cs:12]: Please add some docs over here

File: NCrash/Plugins/ScreenShotWriter.cs [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+        private string _screenshotName;
+        private readonly ILog Logger = LogManager.GetCurrentClassLogger();
+        private string _path;
+
+        
+        /// <summary>
+        /// Generating windows handle list
+        /// </summary>
+
+        internal bool AddWnd(int hwnd, int lparam)
```
Comment (akurdyukov) [inline NCrash/Plugins/ScreenShotWriter.cs]: This may be private

File: NCrash/Plugins/ScreenShotWriter.cs [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+            {
+                Logger.Error("An exception occurred during screenshot generation.", ex);
+            }
+        }
+
+        /// <summary>
+        /// Capturing window area at the screen
+        /// </summary>
+
+        internal Bitmap Capture(int hwnd)
```
Comment (akurdyukov) [inline NCrash/Plugins/ScreenShotWriter.cs]: This may be private

File: NCrash/Plugins/ScreenShotWriter.cs [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+namespace NCrash.Plugins
+{
+    /// <summary>
+    /// Generationg screenshots for current application plugin.
+    /// Add files to AdditionalFiles list
+    /// </summary>
+    public class ScreenShotWriter:IPlugin
+    {
+
+        public ScreenShotWriter()
```
Comment (akurdyukov) [inline NCrash/Plugins/ScreenShotWriter.cs]: You'd better user standard class code structure like
1. consts
2. static
3. readonly fields
4. other fields
5. constructors
6. events
7. properties
8. methods

File: NCrash/ISettings.cs (hunk 14 above)
Comment (traliman) [inline NCrash/ISettings.cs]: If we try to assign this property in any plugin class, we get the compiller error, example:
public void PreProcess(ISettings settings)
        {
....
                if(settings.AdditionalReportFiles == null)
                    settings.AdditionalReportFiles = new List<string>();

Error   1   Property or indexer 'NCrash.ISettings.AdditionalReportFiles' cannot be assigned to -- it is read only

**Pull request merged**